Eliminate scale re-initialization

asfiyab-nvidia opened the pull request:
1. Scale reinitialization fixed my wrapping module init in fp8_autocast and providing recipe
2. Update scale_factors passed to Linear, LayernormLinear, LayernormMLP layers by providing `num_gemms` param and providing 2 scale values per gemm.
3. Add configurable scale initialization for MHA and Transformer Layer

asfiyab-nvidia committed:
Add ONNX export support for TE modules (#1)

* Add TorchScript Operators
* Add symbolic methods to ONNX exporter
* Add tests for the ONNX export

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
fixes for pylint tests

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
fix pylint warning in softmax.py

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
move FP8 ORT lib inside tests/

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
enable cross attention tests

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
refactor code by @nzmora
* Increase layernorm FP16 threshold
* Normalize onnx file names: _ separates configs; - separates words in a single config
* Add get_attn_mask_str and fix mask string
* Add missing ONNX files
* Moved generated ONNX files to tests/gen_onnx_models/

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
fix merge conflict changes

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
fix Q/DQ scale input

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
enable FP16 config when bias is disabled

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
fix scale re-init issues

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
fix pylint check errors

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
update scale arg type in definition

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
add scales to mha and transformer layer submodules

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
updates
1. remove List import for pylint failure
2. address comments: remove state tensors from GPU
3. address comments: Update reverse_map_dtype function and add to namespace

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

asfiyab-nvidia committed:
minor fix: coding guidelines

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>

netaz commented on tests/test_onnx_export.py:
```diff
@@ -80,12 +80,17 @@ def to_numpy(tensor):
-        2, dtype=torch.float32, device="cuda") * scale
+def set_layer_scale(module: torch.nn.Module, scales: float, num_gemms: int=1):
+    module.fp8_init(num_gemms=num_gemms)
+    assert len(scales) == num_gemms * 2, "Each gemm should be accompanied by 2 scales"
```
I don't think we need this assert because it checks a TE predicate that should not affect the export process.

netaz commented on tests/test_onnx_export.py:
```diff
@@ -80,12 +80,17 @@ def to_numpy(tensor):
-        2, dtype=torch.float32, device="cuda") / scale
-    module.fp8_meta["scaling_fwd"].scale_inv = torch.ones(
-        2, dtype=torch.float32, device="cuda") * scale
+def set_layer_scale(module: torch.nn.Module, scales: float, num_gemms: int=1):
```
You can remove `num_gemms` because we don't need/use it.
Please change the typing hint for `scales` to List.

```suggestion
def set_layer_scale(module: torch.nn.Module, scales: List[float]):
```

asfiyab-nvidia replied:
`num_gemms` is required for the `fp8_init()` call. For LayernormMLP specifically, it is set to 2. (see line below)

netaz replied:
My bad - I missed that - ignore my comment

netaz commented on tests/test_onnx_export.py:
```diff
@@ -557,7 +562,7 @@ def forward(self, inp, mask):
 ])
 def test_export_linear(
-    scale_factor: float,
+    scale_factor: list,
```
```suggestion
    scale_factor: List[float],
```

netaz commented on tests/test_onnx_export.py:
```diff
@@ -541,7 +546,7 @@ def forward(self, inp, mask):
 
 
-@pytest.mark.parametrize("scale_factor", [1])
+@pytest.mark.parametrize("scale_factor", [[448, 448]])
```
```suggestion
@pytest.mark.parametrize("scale_factors", [[448, 448]])
```

netaz commented on tests/test_onnx_export.py:
```diff
@@ -610,12 +615,12 @@ def forward(self, inp):
 
 
-@pytest.mark.parametrize("scale_factor", [112])
+@pytest.mark.parametrize("scale_factor", [[448, 448]])
```
```suggestion
@pytest.mark.parametrize("scale_factors", [[448, 448]])
```

Made it explicitly plural

netaz commented on tests/test_onnx_export.py:
```diff
@@ -628,7 +633,7 @@ def forward(self, inp):
 ])
 def test_export_layernorm_linear(
-    scale_factor: float,
+    scale_factor: list,
```
```suggestion
    scale_factor: List[float],
```

netaz commented on tests/test_onnx_export.py:
```diff
@@ -663,7 +668,7 @@ def test_export_layernorm_linear(
 
 
-@pytest.mark.parametrize("scale_factor", [112])
+@pytest.mark.parametrize("scale_factor", [[224, 224, 448, 448]])
```
```suggestion
@pytest.mark.parametrize("scale_factors", [[224, 224, 448, 448]])
```

It will be helpful to add a line documenting which layers these scales affect.

netaz commented on tests/test_onnx_export.py:
```diff
@@ -676,7 +681,7 @@ def test_export_layernorm_linear(
 ])
 def test_export_layernorm_mlp(
-    scale_factor: float,
+    scale_factor: list,
```
```suggestion
    scale_factor: List[float],
```

netaz commented on tests/test_onnx_export.py:
```diff
@@ -704,12 +709,12 @@ def test_export_layernorm_mlp(
         ).to(device='cuda')
         if use_fp8:
-            set_layer_scale(model, scale_factor)
+            set_layer_scale(model, scale_factor, num_gemms=2)
```
```suggestion
            set_layer_scale(model, scale_factors)
```

netaz commented on tests/test_onnx_export.py:
```diff
@@ -772,6 +777,33 @@ def test_export_core_attention(
 
 
+def set_mha_scales(module,
+    scale_factor_qkv: list=[448, 448],
+    scale_factor_query: list=[112, 112],
+    scale_factor_kv: list=[224, 224],
+    scale_factor_proj: list=[448, 448]
```
```suggestion
    scale_factor_qkv: List[float]=[448, 448],
    scale_factor_query: List[float]=[112, 112],
    scale_factor_kv: List[float]=[224, 224],
    scale_factor_proj: List[float]=[448, 448]
```

netaz commented on tests/test_onnx_export.py:
```diff
@@ -805,7 +841,11 @@ def test_export_multihead_attention(
     attention_type: str,
-    fuse_qkv_params: bool
+    fuse_qkv_params: bool,
+    scale_factor_qkv: list,
+    scale_factor_query: list,
+    scale_factor_kv: list,
+    scale_factor_proj: list,
```
```suggestion
    scale_factors_qkv: List[float],
    scale_factors_query: List[float],
    scale_factors_kv: List[float],
    scale_factors_proj: List[float],
```

netaz commented on tests/test_onnx_export.py:
```diff
@@ -846,21 +886,39 @@ def test_export_multihead_attention(
+def set_transformer_layer_scales(module,
+    scales_self_attn: list,
+    scales_inter_attn: list,
+    scales_layernorm_mlp: list=[224, 224, 448, 448]):
```
```suggestion
    scales_layernorm_mlp: List[float]=[224, 224, 448, 448]):
```

netaz commented on tests/test_onnx_export.py:
```diff
@@ -871,14 +929,24 @@ def test_export_multihead_attention(
     fuse_qkv_params: bool,
-    apply_query_key_layer_scaling: bool
+    apply_query_key_layer_scaling: bool,
+    scale_factor_qkv: list,
+    scale_factor_query: list,
+    scale_factor_kv: list,
+    scale_factor_proj: list,
+    scale_factor_layernorm_mlp: list,
```
```suggestion
    scale_factor_qkv: List[float],
    scale_factor_query: List[float],
    scale_factor_kv: List[float],
    scale_factor_proj: List[float],
    scale_factor_layernorm_mlp: List[float],
```

asfiyab-nvidia committed:
address comments
1. replace variable scale_factor with scale_factors
2. Update type hints for scale_factors to be List[float]
3. Remove use of num_gemms param and add amax_history assignment

Signed-off-by: Asfiya Baig <asfiyab@nvidia.com>
